runtime/near-vm-logic/src/logic.rs

```diff
@@ -820,25 +820,24 @@ impl<'a> VMLogic<'a> {
             self.burnt_gas.checked_add(burn_gas as _).ok_or(HostError::IntegerOverflow)?;
         let new_used_gas =
             self.used_gas.checked_add(use_gas as _).ok_or(HostError::IntegerOverflow)?;
-        if self.context.free_of_charge
-            || (new_burnt_gas < self.config.max_gas_burnt
-                && new_used_gas < self.context.prepaid_gas)
+        if new_burnt_gas <= self.config.max_gas_burnt
+            && (self.context.free_of_charge || new_used_gas <= self.context.prepaid_gas)
```

Comment:
Also moving free_of_charge post `max_gas_burnt` limit to avoid infinite loops with view calls.